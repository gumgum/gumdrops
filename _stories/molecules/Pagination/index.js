import React from 'react';
import { number, boolean, select } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';

import readme from './README.md';
import Pagination from '../../../components/molecules/Pagination';

const sizeOptions = ['xl', 'lg', 'sm', 'xs', ''];

class TestPagination extends React.Component {
    state = {
        activePage: 7,
        lastPage: 70
    };

    handlePageChange = newPage => {
        const activePage = newPage;
        this.setState({ activePage });
        action('Page change')(activePage);
    };

    render() {
        const { activePage, lastPage } = this.state;

        return (
            <Pagination
                activePage={number('Current Page', activePage)}
                lastPage={number('Last Page', lastPage)}
                boundaries={boolean('Show Boundaries', false)}
                justify={boolean('Justify', false)}
                size={select('Size', sizeOptions, '')}
                onChange={this.handlePageChange}
            />
        );
    }
}

TestPagination.displayName = 'Pagination';

const component = () => <TestPagination />; // eslint-disable-line

export default [readme, component];
