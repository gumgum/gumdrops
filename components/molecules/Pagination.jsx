import React, { Component } from 'react';
import PropTypes from 'prop-types';
import createRange from '../utils/createRange';
import charCodes from '../../constants/charCodes';
import cx from 'classnames';

class Pagination extends Component {
    _handleKeyPress = event => {
        const { keyCode } = event;

        if (keyCode === charCodes.ARROW_LEFT || keyCode === charCodes.ARROW_UP) {
            this._changePage('prev');
        }
        if (keyCode === charCodes.ARROW_RIGHT || keyCode === charCodes.ARROW_DOWN) {
            this._changePage('next');
        }
    };

    _changePage = page => {
        const { onChange, activePage, lastPage } = this.props;

        const prev = activePage - 1;
        const next = activePage + 1;

        const isFirst = prev < 1;
        const isLast = next > lastPage;

        // Find next page
        let newPage = page;
        if (isNaN(page)) {
            newPage = {
                next: isLast ? lastPage : next,
                prev: isFirst ? 1 : prev
            }[page];
        }

        // Pass next page to callback
        onChange && onChange(newPage);
    };

    _createPages() {
        const { boundaries, activePage, lastPage, justify } = this.props;
        // Beyond this number of results, the indicator doesn't work
        const displayLimit = 7;
        // Number of pages adjacent to the center
        const neighbors = boundaries ? 2 : 3;
        // Set limits on the first section of pages
        const firstSetEnd = boundaries ? displayLimit - 2 : displayLimit - 1;
        // Set limits on the last section of pages
        const lastSetStart = lastPage > displayLimit ? lastPage - firstSetEnd : 1;

        // Generate pages not in first or last sections
        const getMiddleRange = () => {
            const lowerLimit = activePage - neighbors;
            const higherLimit = activePage + neighbors;
            return createRange(lowerLimit, higherLimit);
        };
        // Add first and last pages to current set
        const addBoundaries = (range, type) => {
            const limits = () =>
                ({
                    last: [...range.slice(0, -1), lastPage],
                    first: [1, ...range.slice(-(displayLimit - 1))]
                }[type] || [1, ...range, lastPage]);
            return boundaries ? limits() : range;
        };

        // Total number of pages is below display limit
        const isBelowLimit = lastPage <= displayLimit;
        // Current page is within first section
        const isFirstSet = !isBelowLimit && activePage <= firstSetEnd;
        // Current page is within last section
        const isLastSet = !isFirstSet && activePage > lastSetStart;

        // Set of pages when total is below display limit
        const shortSet = isBelowLimit && createRange(1, lastPage);
        // First set of pages
        const firstSet = isFirstSet && addBoundaries(createRange(1, displayLimit), 'last');
        // Last set of pages
        const lastSet = isLastSet && addBoundaries(createRange(lastSetStart, lastPage), 'first');

        const currentSet = shortSet || firstSet || lastSet || addBoundaries(getMiddleRange());

        return currentSet.map(page => {
            const isCurrent = page === activePage;
            const eltClass = cx('gds-pagination__page-item', '-va-top', {
                'gds-pagination__page-item--fixed': !justify,
                'gds-pagination__page-item--active': isCurrent
            });
            const callback = () => this._changePage(page);
            return { page, eltClass, callback, isCurrent };
        });
    }

    render() {
        const { className, justify, size, lastPage, activePage } = this.props;

        const isPrevDisabled = activePage === 1;
        const isNextDisabled = activePage === lastPage;

        const rootClass = cx('gds-pagination', 'gds-pagination--mobile-arrows', className, {
            'gds-pagination--fixed': !justify,
            [`gds-pagination--${size}`]: size
        });

        const itemClass = cx('gds-pagination__page-item', '-va-top', {
            'gds-pagination__page-item--fixed': !justify
        });

        const linkClass = cx('gds-pagination__page-link', '-overflow-hidden', {
            'gds-pagination__page-link--fixed': !justify
        });

        // fixes style issues associated with using buttons
        const btnStyle = {
            border: 0,
            fontFamily: 'inherit',
            padding: justify ? null : 0,
            background: 'none',
            height: size === 'xs' ? '1.5rem' : ''
        };

        const shouldRender = lastPage > 1 && activePage > 0 && activePage <= lastPage;
        if (!shouldRender) return null;

        return (
            <nav className={rootClass} onKeyDown={this._handleKeyPress}>
                <div className={itemClass}>
                    <button
                        name="prev"
                        aria-label="Goto previous page"
                        className={cx(linkClass, { '-disabled': isPrevDisabled })}
                        disabled={isPrevDisabled}
                        onClick={() => this._changePage('prev')}
                        style={btnStyle}>
                        <span className="-vis-hidden -ellipsis">Go to previous page</span>
                    </button>
                </div>
                {this._createPages().map(({ page, eltClass, callback, isCurrent }) => (
                    <div key={page} className={eltClass}>
                        <button
                            aria-label={isCurrent ? `Page ${page}, Current` : `Goto page ${page}`}
                            aria-current={isCurrent}
                            className={linkClass}
                            onClick={callback}
                            style={btnStyle}>
                            {page}
                        </button>
                    </div>
                ))}
                <div className={itemClass}>
                    <button
                        name="next"
                        aria-label="Goto next page"
                        className={cx(linkClass, { '-disabled': isNextDisabled })}
                        disabled={isNextDisabled}
                        onClick={() => this._changePage('next')}
                        style={btnStyle}>
                        <span className="-vis-hidden -ellipsis">Go to next page</span>
                    </button>
                </div>
                <span className="gds-pagination__page-indicator" />
            </nav>
        );
    }
}

Pagination.displayName = 'Pagination';

Pagination.propTypes = {
    onChange: PropTypes.func.isRequired,
    lastPage: PropTypes.number.isRequired,
    activePage: PropTypes.number,
    boundaries: PropTypes.bool,
    justify: PropTypes.bool,
    size: PropTypes.string,
    className: PropTypes.string
};

Pagination.defaultProps = {
    activePage: 1,
    itemsPerPage: 10,
    boundaries: false,
    justify: false,
    className: ''
};

export default Pagination;
