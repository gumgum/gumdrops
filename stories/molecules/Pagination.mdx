import { Story, Props } from '@storybook/addon-docs/blocks';
import Pagination from '../../components/molecules/Pagination';

# Pagination

The `<Pagination>` component creates a set of indicators to show how many pages exist for the current list and helps navigate them.

When a page change is detected, the `onChange` prop will be called with the next page number as its only argument.

## Full Example

```jsx
const [activePage, setActivePage] = useState(1);
return (
    <Pagination
        activePage={activePage}
        lastPage={30}
        onChange={nextPage => setActivePage(nextPage)}
    />
);
```

## Keyboard Accessibility

When pagination is in focus, you can use the arrow key to navigate to the next (arrow right and arrow down) and previous (arrow up and arrow left) pages. You can also focus on the next/previous button and "press" them with the spacebar or enter keys.

## Props

<Props of={Pagination} />
