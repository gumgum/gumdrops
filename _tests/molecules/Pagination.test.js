/* globals mount */
import React from 'react';
import Pagination from '../../components/molecules/Pagination';

const defaultProps = {
    onChange: () => {},
    lastPage: 100,
    activePage: 6,
    boundaries: false,
    justify: false,
    size: 'sm',
    className: 'custom-class'
};

describe('Expect <Pagination>', () => {
    it('to render', () => {
        const wrapper = mount(<Pagination {...defaultProps} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('to render with boundaries and justify', () => {
        const props = {
            ...defaultProps,
            boundaries: true,
            justify: true
        };
        const wrapper = mount(<Pagination {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('to disable next/prev buttons', () => {
        const props = {
            onChange: () => {},
            lastPage: 2,
            activePage: 1,
            boundaries: true,
            justify: true
        };
        const wrapper = mount(<Pagination {...props} />);
        expect(wrapper.find('button [name="prev"]').props().disabled).toBe(true);
        expect(wrapper.find('button [name="next"]').props().disabled).toBe(false);
        wrapper.setProps({ activePage: 2 });
        expect(wrapper.find('button [name="prev"]').props().disabled).toBe(false);
        expect(wrapper.find('button [name="next"]').props().disabled).toBe(true);
    });
});
