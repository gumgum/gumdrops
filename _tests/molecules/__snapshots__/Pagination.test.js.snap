// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <Pagination> to render 1`] = `
<Pagination
  activePage={6}
  boundaries={false}
  className="custom-class"
  itemsPerPage={10}
  justify={false}
  lastPage={100}
  onChange={[Function]}
  size="sm"
>
  <nav
    className="gds-pagination gds-pagination--mobile-arrows custom-class gds-pagination--fixed gds-pagination--sm"
    onKeyDown={[Function]}
  >
    <div
      className="gds-pagination__page-item -va-top gds-pagination__page-item--fixed"
    >
      <button
        aria-label="Goto previous page"
        className="gds-pagination__page-link -overflow-hidden gds-pagination__page-link--fixed"
        disabled={false}
        name="prev"
        onClick={[Function]}
        style={
          Object {
            "background": "none",
            "border": 0,
            "fontFamily": "inherit",
            "height": "",
            "padding": 0,
          }
        }
      >
        <span
          className="-vis-hidden -ellipsis"
        >
          Go to previous page
        </span>
      </button>
    </div>
    <div
      className="gds-pagination__page-item -va-top gds-pagination__page-item--fixed"
      key="1"
    >
      <button
        aria-current={false}
        aria-label="Goto page 1"
        className="gds-pagination__page-link -overflow-hidden gds-pagination__page-link--fixed"
        onClick={[Function]}
        style={
          Object {
            "background": "none",
            "border": 0,
            "fontFamily": "inherit",
            "height": "",
            "padding": 0,
          }
        }
      >
        1
      </button>
    </div>
    <div
      className="gds-pagination__page-item -va-top gds-pagination__page-item--fixed"
      key="2"
    >
      <button
        aria-current={false}
        aria-label="Goto page 2"
        className="gds-pagination__page-link -overflow-hidden gds-pagination__page-link--fixed"
        onClick={[Function]}
        style={
          Object {
            "background": "none",
            "border": 0,
            "fontFamily": "inherit",
            "height": "",
            "padding": 0,
          }
        }
      >
        2
      </button>
    </div>
    <div
      className="gds-pagination__page-item -va-top gds-pagination__page-item--fixed"
      key="3"
    >
      <button
        aria-current={false}
        aria-label="Goto page 3"
        className="gds-pagination__page-link -overflow-hidden gds-pagination__page-link--fixed"
        onClick={[Function]}
        style={
          Object {
            "background": "none",
            "border": 0,
            "fontFamily": "inherit",
            "height": "",
            "padding": 0,
          }
        }
      >
        3
      </button>
    </div>
    <div
      className="gds-pagination__page-item -va-top gds-pagination__page-item--fixed"
      key="4"
    >
      <button
        aria-current={false}
        aria-label="Goto page 4"
        className="gds-pagination__page-link -overflow-hidden gds-pagination__page-link--fixed"
        onClick={[Function]}
        style={
          Object {
            "background": "none",
            "border": 0,
            "fontFamily": "inherit",
            "height": "",
            "padding": 0,
          }
        }
      >
        4
      </button>
    </div>
    <div
      className="gds-pagination__page-item -va-top gds-pagination__page-item--fixed"
      key="5"
    >
      <button
        aria-current={false}
        aria-label="Goto page 5"
        className="gds-pagination__page-link -overflow-hidden gds-pagination__page-link--fixed"
        onClick={[Function]}
        style={
          Object {
            "background": "none",
            "border": 0,
            "fontFamily": "inherit",
            "height": "",
            "padding": 0,
          }
        }
      >
        5
      </button>
    </div>
    <div
      className="gds-pagination__page-item -va-top gds-pagination__page-item--fixed gds-pagination__page-item--active"
      key="6"
    >
      <button
        aria-current={true}
        aria-label="Page 6, Current"
        className="gds-pagination__page-link -overflow-hidden gds-pagination__page-link--fixed"
        onClick={[Function]}
        style={
          Object {
            "background": "none",
            "border": 0,
            "fontFamily": "inherit",
            "height": "",
            "padding": 0,
          }
        }
      >
        6
      </button>
    </div>
    <div
      className="gds-pagination__page-item -va-top gds-pagination__page-item--fixed"
      key="7"
    >
      <button
        aria-current={false}
        aria-label="Goto page 7"
        className="gds-pagination__page-link -overflow-hidden gds-pagination__page-link--fixed"
        onClick={[Function]}
        style={
          Object {
            "background": "none",
            "border": 0,
            "fontFamily": "inherit",
            "height": "",
            "padding": 0,
          }
        }
      >
        7
      </button>
    </div>
    <div
      className="gds-pagination__page-item -va-top gds-pagination__page-item--fixed"
    >
      <button
        aria-label="Goto next page"
        className="gds-pagination__page-link -overflow-hidden gds-pagination__page-link--fixed"
        disabled={false}
        name="next"
        onClick={[Function]}
        style={
          Object {
            "background": "none",
            "border": 0,
            "fontFamily": "inherit",
            "height": "",
            "padding": 0,
          }
        }
      >
        <span
          className="-vis-hidden -ellipsis"
        >
          Go to next page
        </span>
      </button>
    </div>
    <span
      className="gds-pagination__page-indicator"
    />
  </nav>
</Pagination>
`;

exports[`Expect <Pagination> to render with boundaries and justify 1`] = `
<Pagination
  activePage={6}
  boundaries={true}
  className="custom-class"
  itemsPerPage={10}
  justify={true}
  lastPage={100}
  onChange={[Function]}
  size="sm"
>
  <nav
    className="gds-pagination gds-pagination--mobile-arrows custom-class gds-pagination--sm"
    onKeyDown={[Function]}
  >
    <div
      className="gds-pagination__page-item -va-top"
    >
      <button
        aria-label="Goto previous page"
        className="gds-pagination__page-link -overflow-hidden"
        disabled={false}
        name="prev"
        onClick={[Function]}
        style={
          Object {
            "background": "none",
            "border": 0,
            "fontFamily": "inherit",
            "height": "",
            "padding": null,
          }
        }
      >
        <span
          className="-vis-hidden -ellipsis"
        >
          Go to previous page
        </span>
      </button>
    </div>
    <div
      className="gds-pagination__page-item -va-top"
      key="1"
    >
      <button
        aria-current={false}
        aria-label="Goto page 1"
        className="gds-pagination__page-link -overflow-hidden"
        onClick={[Function]}
        style={
          Object {
            "background": "none",
            "border": 0,
            "fontFamily": "inherit",
            "height": "",
            "padding": null,
          }
        }
      >
        1
      </button>
    </div>
    <div
      className="gds-pagination__page-item -va-top"
      key="4"
    >
      <button
        aria-current={false}
        aria-label="Goto page 4"
        className="gds-pagination__page-link -overflow-hidden"
        onClick={[Function]}
        style={
          Object {
            "background": "none",
            "border": 0,
            "fontFamily": "inherit",
            "height": "",
            "padding": null,
          }
        }
      >
        4
      </button>
    </div>
    <div
      className="gds-pagination__page-item -va-top"
      key="5"
    >
      <button
        aria-current={false}
        aria-label="Goto page 5"
        className="gds-pagination__page-link -overflow-hidden"
        onClick={[Function]}
        style={
          Object {
            "background": "none",
            "border": 0,
            "fontFamily": "inherit",
            "height": "",
            "padding": null,
          }
        }
      >
        5
      </button>
    </div>
    <div
      className="gds-pagination__page-item -va-top gds-pagination__page-item--active"
      key="6"
    >
      <button
        aria-current={true}
        aria-label="Page 6, Current"
        className="gds-pagination__page-link -overflow-hidden"
        onClick={[Function]}
        style={
          Object {
            "background": "none",
            "border": 0,
            "fontFamily": "inherit",
            "height": "",
            "padding": null,
          }
        }
      >
        6
      </button>
    </div>
    <div
      className="gds-pagination__page-item -va-top"
      key="7"
    >
      <button
        aria-current={false}
        aria-label="Goto page 7"
        className="gds-pagination__page-link -overflow-hidden"
        onClick={[Function]}
        style={
          Object {
            "background": "none",
            "border": 0,
            "fontFamily": "inherit",
            "height": "",
            "padding": null,
          }
        }
      >
        7
      </button>
    </div>
    <div
      className="gds-pagination__page-item -va-top"
      key="8"
    >
      <button
        aria-current={false}
        aria-label="Goto page 8"
        className="gds-pagination__page-link -overflow-hidden"
        onClick={[Function]}
        style={
          Object {
            "background": "none",
            "border": 0,
            "fontFamily": "inherit",
            "height": "",
            "padding": null,
          }
        }
      >
        8
      </button>
    </div>
    <div
      className="gds-pagination__page-item -va-top"
      key="100"
    >
      <button
        aria-current={false}
        aria-label="Goto page 100"
        className="gds-pagination__page-link -overflow-hidden"
        onClick={[Function]}
        style={
          Object {
            "background": "none",
            "border": 0,
            "fontFamily": "inherit",
            "height": "",
            "padding": null,
          }
        }
      >
        100
      </button>
    </div>
    <div
      className="gds-pagination__page-item -va-top"
    >
      <button
        aria-label="Goto next page"
        className="gds-pagination__page-link -overflow-hidden"
        disabled={false}
        name="next"
        onClick={[Function]}
        style={
          Object {
            "background": "none",
            "border": 0,
            "fontFamily": "inherit",
            "height": "",
            "padding": null,
          }
        }
      >
        <span
          className="-vis-hidden -ellipsis"
        >
          Go to next page
        </span>
      </button>
    </div>
    <span
      className="gds-pagination__page-indicator"
    />
  </nav>
</Pagination>
`;
